src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1019,29 +1063,38 @@ public SolutionCompilationState WithoutFrozenSourceGeneratedDocuments()
             return this;
 
         var documentStatesByProjectId = documentStates.ToDictionary(static state => state.Id.ProjectId);
-        var newTrackerMap = CreateCompilationTrackerMap(documentStatesByProjectId.Keys.ToImmutableArray(), this.SolutionState.GetProjectDependencyGraph());
-
-        foreach (var (projectId, documentStatesForProject) in documentStatesByProjectId)
-        {
-            // We want to create a new snapshot with a new compilation tracker that will do this replacement.
-            // If we already have an existing tracker we'll just wrap that (so we also are reusing any underlying
-            // computations). If we don't have one, we'll create one and then wrap it.
-            if (!newTrackerMap.TryGetValue(projectId, out var existingTracker))
-            {
-                existingTracker = CreateCompilationTracker(projectId, this.SolutionState);
-            }
-
-            newTrackerMap = newTrackerMap.SetItem(
-                projectId,
-                new GeneratedFileReplacingCompilationTracker(existingTracker, new(documentStatesForProject)));
-        }
+        var newTrackerMap = CreateCompilationTrackerMap(
+            documentStatesByProjectId.Keys.ToImmutableArray(),
+            this.SolutionState.GetProjectDependencyGraph(),
+            modifyTrackerMapInfo,
+            (documentStatesByProjectId, this.SolutionState));
 
         // We pass the same solution state, since this change is only a change of the generated documents -- none of the core
         // documents or project structure changes in any way.
         return this.Branch(
             this.SolutionState,
             projectIdToTrackerMap: newTrackerMap,
             frozenSourceGeneratedDocumentStates: new TextDocumentStates<SourceGeneratedDocumentState>(documentStates));
+
+        static bool modifyTrackerMapInfo(
+            Dictionary<ProjectId, ICompilationTracker> trackerMap,
+            (Dictionary<ProjectId, ImmutableArray<SourceGeneratedDocumentState>> documentStatesByProjectId, SolutionState solutionState) arg)
+        {
+            foreach (var (projectId, documentStatesForProject) in arg.documentStatesByProjectId)
+            {
+                // We want to create a new snapshot with a new compilation tracker that will do this replacement.
+                // If we already have an existing tracker we'll just wrap that (so we also are reusing any underlying
+                // computations). If we don't have one, we'll create one and then wrap it.
+                if (!trackerMap.TryGetValue(projectId, out var existingTracker))
+                {
+                    existingTracker = CreateCompilationTracker(projectId, arg.solutionState);
+                }
+
+                trackerMap[projectId] = new GeneratedFileReplacingCompilationTracker(existingTracker, new(documentStatesForProject));
+            }
+
+            return true;
```

Comment:
same question about returning true.  or do we think it just doesn't matter in these cases?

Reply:
I didn't think the non-changed case was likely in either of these too, but I went ahead and changed it.